components/support/nimbus-fml/src/util/loaders.rs

```diff
@@ -0,0 +1,289 @@
+        self.cache_dir.join(filename)
+    }
+
+    /// Joins a path to a string, to make a new path.
+    ///
+    /// We want to be able to support local and remote files.
+    /// We also want to be able to support a configurable short cut format.
+    /// Following a pattern common in other package managers, `@XXXX/YYYY`
+    /// is used as short hand for the main branch in github repos.
+    pub fn join(&self, base: &FilePath, f: &str) -> Result<FilePath> {
+        if f.starts_with('@') {
+            let f = f.replacen('@', "", 1);
+            let parts = f.splitn(3, '/').collect::<Vec<&str>>();
+            match parts.as_slice() {
+                [user, repo, path] => {
+                    let repo = self.lookup_repo_path(user, repo);
+                    let f = format!("{}/{}", repo, path);
+                    base.join(&f)
+                }
+                _ => Err(FMLError::InvalidPath(format!(
+                    "'{}' needs to include a username, a repo and a filepath",
+                    f
+                ))),
+            }
+        } else {
+            base.join(f)
+        }
+    }
+
+    fn lookup_repo_path(&self, user: &str, repo: &str) -> String {
+        let default = "main".to_string();
+
+        let value = self
+            .config
+            .get(&format!("{}/{}", user, repo))
+            .or_else(|| self.config.get(&format!("@{}/{}", user, repo)))
+            .unwrap_or(&default);
+
+        if value.contains("://") || value.starts_with('.') {
+            // It's a URL, or a relative path!
+            value.clone()
+        } else {
+            // It's a branch.
+            let branch = value;
+            [GITHUB_USER_CONTENT_DOTCOM, user, repo, branch].join("/")
+        }
+    }
+}
+
+#[cfg(test)]
+mod unit_tests {
+    use super::*;
+
+    #[test]
+    fn test_relative_paths() -> Result<()> {
+        let tmp = std::env::temp_dir();
+
+        let file = tmp.join("foo/bar.txt");
+        let obs = FilePath::from(file.as_path());
+
+        assert!(matches!(obs, FilePath::Local(_)));
+        assert!(obs.to_string().ends_with("foo/bar.txt"));
+
+        let obs = obs.join("baz.txt")?;
+        assert!(obs.to_string().ends_with("foo/baz.txt"));
+
+        let obs = obs.join("./bam.txt")?;
+        // We'd prefer it to be like this:
+        // assert!(obs.to_string().ends_with("foo/bam.txt"));
+        // But there's no easy way to get this (because symlinks).
+        // This is most likely the correct thing for us to do.
+        // We put this test here for documentation purposes, and to
+        // highlight that with URLs, ../ and ./ do what youj might
+        // expect.
+        assert!(obs.to_string().ends_with("foo/./bam.txt"));
+
+        let obs = obs.join("https://example.com/foo/bar.txt")?;
+        assert!(matches!(obs, FilePath::Remote(_)));
+        assert!(obs.to_string().ends_with("example.com/foo/bar.txt"));
```

Comment:
should we be a bit more strict and test that the whole string is the URL? (since it was a local path, but is being joined by a URL)